<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Workflow &amp; Calibration</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Home</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Cal_ChIP-seq_workshop_intro.html">Getting Started</a>
</li>
<li>
  <a href="Cal_ChIP-seq_workshop_QC.html">Quality Control</a>
</li>
<li>
  <a href="Cal_ChIP-seq_workshop_map.html">Mapping and Filtering</a>
</li>
<li>
  <a href="Cal_ChIP-seq_workshop_visualisation.html">Visualisation</a>
</li>
<li>
  <a href="Cal_ChIP-seq_workshop_cal.html">Calibration</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Workflow &amp; Calibration</h1>

</div>


<head>
<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>
</head>
<p><br></p>
<div id="workflow-calibration" class="section level3">
<h3>9. Workflow &amp; Calibration</h3>
<p>Here is a bash script which shows all commands necessary to get to the calibration stage, you should recognise these steps. You’ll notice output is organized into specific directories, this is useful when working with multiple samples.</p>
<div class="blue">
<pre class="bash"><code>### Need to set
#PATH/calibration_genome_reference.fa
#PATH/experimental_genome_reference.fa
#PATH/calibration_genome_reference.chrom.sizes
#PATH/experimental_genome_reference.chrom.sizes
###

### First mapping to **Calibration** Genome

mkdir miniMap2_out_calibration_genome

# Map raw/trimmed reads to calibration genome with Minimap2 (paired end)
minimap2 -ax sr PATH/calibration_genome_reference.fa cutadapt/$1.trimmed.fastq.gz cutadapt/$2.trimmed.fastq.gz &gt; miniMap2_out_calibration_genome/$3.calibration.mm2.sam

# SAM to BAM to sort (unmapped reads only)
samtools view -Sbh -f 4 miniMap2_out_calibration_genome/$3.calibration.mm2.sam -o miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.bam

# sort unmapped calibration BAM (-n by qname)
samtools sort -n -@ 5 miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.bam -o miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.qsort.bam

# interleaved (keep in single fastq)
samtools bam2fq miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.qsort.bam &gt; miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.fastq
gzip miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.fastq

# clean up 
rm miniMap2_out_calibration_genome/$3.calibration.mm2.sam
rm miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.bam 
rm miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.qsort.bam


### First mapping to **Experimental** Genome

mkdir miniMap2_out_experimental_genome

# Map raw/trimmed reads to experimental genome with Minimap2 (paired end)
minimap2 -ax sr PATH/experimental_genome_reference.fa cutadapt/$1.trimmed.fastq.gz cutadapt/$2.trimmed.fastq.gz &gt; miniMap2_out_experimental_genome/$3.experimental.mm2.sam

# SAM to BAM to sort (unmapped reads only)
samtools view -Sbh -f 4 miniMap2_out_experimental_genome/$3.experimental.mm2.sam -o miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.bam

# sort unmapped sacer BAM (-n by qname)
samtools sort -n -@ 5 miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.bam -o miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.qsort.bam

# interleaved (keep in single fastq)
samtools bam2fq miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.qsort.bam &gt; miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.fastq
gzip miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.fastq

# clean up
rm miniMap2_out_experimental_genome/$3.experimental.mm2.sam
rm miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.bam
rm miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.qsort.bam 


### Second mapping to **Calibration** Genome

mkdir miniMap2_out_calibration_genome_only

# Map unmapped experimental genome reads to calibration genome with Minimap2 (interleaved)
# input miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.fastq.gz
minimap2 -ax sr PATH/calibration_genome_reference.fa miniMap2_out_experimental_genome/$3.experimental.mm2.unmapped.fastq.gz &gt; miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.sam

# SAM to BAM to sort (mapped reads only)
samtools view -Sbh -F 4 miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.sam -o miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.bam
samtools sort -@ 5 miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.bam -o miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.sorted.bam

rm miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.bam
rm miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.sam
mv miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.sorted.bam miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.bam

samtools index miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.bam

# remove chrMT/MTR/AB
samtools view -b miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.bam I II III &gt; miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.chrMT.bam
samtools index miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.chrMT.bam
samtools flagstat miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.chrMT.bam &gt; miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.chrMT.bam.fs

###
# Create BigWigs (READS PER MILLION)
# Get the number of mapped reads from flagstat
num1=`cat  miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.chrMT.bam.fs | grep N/A | grep mapped | grep N/A |cut -f1 -d &#39; &#39;`
echo &quot;$3: $num1&quot;

# Convert to RPM
RPM=`perl -e &quot;print 1000000/$num1&quot;`
echo &quot;The scaling factor is $RPM&quot;

# Create a bedgraph file scaled to RPM 
genomeCoverageBed -ibam  miniMap2_out_calibration_genome_only/$3.calibration.only.mm2.chrMT.bam -bg -scale $RPM &gt; miniMap2_out_calibration_genome_only/$3.calibration.only.chrMT.rpm.bedgraph

# Convert to bigwig 
wigToBigWig miniMap2_out_calibration_genome_only/$3.calibration.only.chrMT.rpm.bedgraph PATH/calibration_genome_reference.chrom.sizes miniMap2_out_calibration_genome_only/$3.calibration.only.chrMT.rpm.bw


### Second mapping to **Experimental** Genome

mkdir miniMap2_out_experimental_genome_only

# Map unmapped calibration genome to experimental genome with Minimap2 (interleaved)
# input miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.fastq
minimap2 -ax sr PATH/experimental_genome_reference.fa miniMap2_out_calibration_genome/$3.calibration.mm2.unmapped.fastq.gz &gt; miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.sam

# SAM to BAM to sort (mapped reads only)
samtools view -Sbh -F 260 -f 3 miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.sam -o miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.bam
samtools sort -@ 5 miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.bam -o miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.sorted.bam

rm miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.bam
rm miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.sam
mv miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.sorted.bam miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.bam

samtools index miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.bam

# exclude chrMT
samtools view -b miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.bam chrI chrII chrIII chrIV chrV chrVI chrVII chrVIII chrIX chrX chrXI chrXII chrXIII chrXIV chrXV chrXVI &gt; miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.bam

samtools index miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.bam
samtools flagstat miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.bam &gt; miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.bam.fs

# Remove rDNA regions (chrXII saturated with reads)
bedtools intersect -v -abam miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.bam -b /homes/genomes/training/calibration_workshop/rDNA.bed &gt; miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.rDNA.bam
samtools index miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.rDNA.bam

###
# Create BigWigs (READS PER MILLION)
# Get the number of mapped reads from flagstat
num1=`cat  miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.rDNA.bam.fs | grep N/A|grep mapped | grep N/A |cut -f1 -d &#39; &#39;`
echo &quot;$3: $num1&quot;

# Convert to RPM
RPM=`perl -e &quot;print 1000000/$num1&quot;`
echo &quot;The scaling factor is $RPM&quot;

# Create a bedgraph file scaled to RPM
genomeCoverageBed -ibam  miniMap2_out_experimental_genome_only/$3.experimental.only.mm2.MT.rDNA.bam -bg -scale $RPM  &gt; miniMap2_out_experimental_genome_only/$3.experimental.only.MT.rDNA.rpm.bedgraph

# Convert to bigwig
wigToBigWig  miniMap2_out_experimental_genome_only/$3.experimental.only.MT.rDNA.rpm.bedgraph PATH/experimental_genome_reference.chrom.sizes miniMap2_out_experimental_genome_only/$3.experimental.only.MT.rDNA.rpm.bw</code></pre>
</div>
<p>Once we have mapped INPUT and IP reads to each genome, we can perform the calibration. This is similiar to how we created the RPM tracks;</p>
<div class="blue">
<pre class="bash"><code>mkdir calibrate

# calculate OR value
# Get the number of mapped reads from flagstat

# calibration input
num1=`cat  miniMap2_out_calibration_genome_only/$1.calibration.only.mm2.chrMT.bam.fs | grep N/A|grep mapped | grep N/A |cut -f1 -d &#39; &#39;`
echo &quot;$1 pombe: $num1 (Wc)&quot;

# experimental IP
num2=`cat  miniMap2_out_experimental_genome_only/$2.experimental.only.mm2.MT.rDNA.bam.fs | grep N/A|grep mapped | grep N/A |cut -f1 -d &#39; &#39;`
echo &quot;$2 sacCer3: $num2 (IPx)&quot;

# calibration IP
num3=`cat  miniMap2_out_calibration_genome_only/$2.calibration.only.mm2.chrMT.bam.fs | grep N/A|grep mapped | grep N/A |cut -f1 -d &#39; &#39;`
echo &quot;$2 pombe: $num3 (IPc)&quot;

# experimental input
num4=`cat  miniMap2_out_experimental_genome_only/$1.experimental.only.mm2.MT.rDNA.bam.fs | grep N/A|grep mapped | grep N/A |cut -f1 -d &#39; &#39;`
echo &quot;$1 sacCer3: $num4 (Wx)&quot;

WcIPx=`perl -e &quot;print $num1*$num2&quot;`
WxIPc=`perl -e &quot;print $num4*$num3&quot;`

OR=`perl -e &quot;print $WcIPx / $WxIPc;&quot;`
echo &quot;The occupancy ratio is $OR&quot;

# Create BigWigs (Calibrated)
cat miniMap2_out_experimental_genome_only/$2.experimental.only.MT.rDNA.rpm.bedgraph | awk &#39;{print $1 &quot;\t&quot; $2 &quot;\t&quot; $3 &quot;\t&quot; $4*&#39;$OR&#39;}&#39; &gt; calibrate/$2.experimental.only.MT.rDNA.rpm.calibrated.bedgraph

# Convert to bigwig
wigToBigWig calibrate/$2.experimental.only.MT.rDNA.rpm.calibrated.bedgraph PATH_TO/Reference.chrom.sizes calibrate/$2.experimental.only.MT.rDNA.rpm.calibrated.bw</code></pre>
</div>
<p>Rather than running a simple bash script we have setup a Snakemake workflow that will run all of our samples through the full mapping and calibration process.</p>
Lets create a new directory for this and get things setup.
<div class="blue">
<pre class="bash"><code>cd
mkdir calibration_workflow
cd calibration_workflow
cp /library/pipelines/snakemake/marston_ChIP-seq_PE_calibration/v1/* . </code></pre>
</div>
<p>To run the Snakemake workflow</p>
<p>It’s a good idea to first do a ‘dryrun’ which won’t actually execute the commands, but makes sure everything is in order.</p>
<div class="blue">
<pre class="bash"><code>snakemake -npr --cores 5</code></pre>
</div>
<p>-n = dryrun</p>
<p>-p = print</p>
<p>-r = reason</p>
Now, lets run the workflow
<div class="blue">
<pre class="bash"><code>snakemake -pr --cores 5</code></pre>
<p>This will take few mintues to to finish, so lets go back to the IGV demo.</p>
<p>Once this is complete, see if you can link the new QC reports/bigWigs to your public_html directory.</p>
<p><br></p>
</div>
<div id="snakemake" class="section level3">
<h3>Snakemake</h3>
<p>There are several workflow managment systems, Snakemake is something we are using increasingly. <a href="https://snakemake.readthedocs.io/en/stable/" class="uri">https://snakemake.readthedocs.io/en/stable/</a></p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
